extensions/aws/processors/ListS3.cpp

```diff
@@ -0,0 +1,295 @@
+  auto it = state.find(LATEST_LISTED_KEY_TIMESTAMP);
+  if (it != state.end()) {
+    stored_listed_key_timestamp_str = it->second;
+  }
+
+  int64_t stored_listed_key_timestamp = 0;
+  core::Property::StringToInt(stored_listed_key_timestamp_str, stored_listed_key_timestamp);
+
+  return stored_listed_key_timestamp;
+}
+
+ListS3::ListingState ListS3::getCurrentState(const std::shared_ptr<core::ProcessContext> &context) {
+  ListS3::ListingState current_listing_state;
+  std::unordered_map<std::string, std::string> state;
+  if (!state_manager_->get(state)) {
+    logger_->log_info("No stored state for listed objects was found");
+    return current_listing_state;
+  }
+
+  current_listing_state.listed_key_timestamp = getLatestListedKeyTimestamp(state);
+  logger_->log_debug("Restored previous listed timestamp %lld", current_listing_state.listed_key_timestamp);
+
+  current_listing_state.listed_keys = getLatestListedKeys(state);
+  return current_listing_state;
+}
+
+void ListS3::storeState(const std::shared_ptr<core::ProcessContext> &context, const ListS3::ListingState &latest_listing_state) {
+  std::unordered_map<std::string, std::string> state;
+  state[LATEST_LISTED_KEY_TIMESTAMP] = std::to_string(latest_listing_state.listed_key_timestamp);
+  for (std::size_t i = 0; i < latest_listing_state.listed_keys.size(); ++i) {
+    state[LATEST_LISTED_KEY_PREFIX + std::to_string(i)] = latest_listing_state.listed_keys.at(i);
+  }
+  logger_->log_debug("Stored new listed timestamp %lld", latest_listing_state.listed_key_timestamp);
+  state_manager_->set(state);
+}
+
+void ListS3::createNewFlowFile(
+    const std::shared_ptr<core::ProcessSession> &session,
+    const aws::s3::ListedObjectAttributes &object_attributes) {
+  auto flow_file = session->create();
+  session->putAttribute(flow_file, "s3.bucket", list_request_params_.bucket);
+  session->putAttribute(flow_file, core::SpecialFlowAttribute::FILENAME, object_attributes.filename);
+  session->putAttribute(flow_file, "s3.etag", object_attributes.etag);
+  session->putAttribute(flow_file, "s3.isLatest", object_attributes.is_latest ? "true" : "false");
+  session->putAttribute(flow_file, "s3.lastModified", std::to_string(object_attributes.last_modified));
+  session->putAttribute(flow_file, "s3.length", std::to_string(object_attributes.length));
+  session->putAttribute(flow_file, "s3.storeClass", object_attributes.store_class);
+  if (!object_attributes.version.empty()) {
+    session->putAttribute(flow_file, "s3.version", object_attributes.version);
+  }
+  writeObjectTags(list_request_params_.bucket, object_attributes, session, flow_file);
+  writeUserMetadata(object_attributes, session, flow_file);
+
+  session->transfer(flow_file, Success);
+}
+
+void ListS3::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  logger_->log_debug("ListS3 onTrigger");
+
+  auto aws_results = s3_wrapper_.listBucket(list_request_params_);
```

Comment:
would it be possible to use the previous `onTrigger`s timestamp to update the minimum object age and avoid refetching the whole object list each time?

Reply:
Unfortunately the AWS SDK client does not provide the minimum object age as a parameter so we still have to do the filtering by hand either in the `onTrigger` or in the `S3Wrapper`, so all of the objects will be retrieved at all times. Even if we would pass the new timestamp to the `S3Wrapper` the objects still need to be filtered in the `onTrigger` as well to see if there were additional objects fetched with that last timestamp since the previous listing. I would not mix these two filters, I think it's a bit more easier to follow the logic if we have a separate filter with the user given minimum object age in the `S3Wrapper`, and have another filter in the `onTrigger` with the parameters retrieved from our last stored state.

Reply:
thank you, makes sense